### PR TITLE
Update Bazel & protobufs, then narrow warnings to Carbon.

### PR DESCRIPTION
**Pull request by chandlerc**
Protobufs code hits a warning with the latest system headers on macOS. I figured this may have been fixed so I updated protobufs and Bazel to the latest releases. This generally cleaned things up.

However, it actually added *more* warnings. This clearly isn't a really well tested path. In fact, we already have a disabled warning that we'd like for Carbon code because LLVM isn't clean for that warning.

So I've switched our warning strategy to a more durable approach of suppressing all warnings for external repository headers and source files. This lets us re-enable the missing warning and should fix the protobuf warning that started me down this twisty path.

Sadly, we *have* to update to Bazel 6 in order to have the necessary flag to use this approach to suppressing warnings, so I couldn't do this as two PRs cleanly. =/ That's why I've bundled both the Bazel (and protobuf) updates with the warning strategy change.

Last but not least, I've fixed several unused parameters in Carbon's code that our warnings now catch.

### CODE REVIEW

File: bazel/cc_toolchains/clang_cc_toolchain_config.bzl [hunk 1]
```diff
@@ -133,8 +133,9 @@ def _impl(ctx):
                             "-Wimplicit-fallthrough",
                             "-Wctad-maybe-unsupported",
                             "-Wnon-virtual-dtor",
-                            # Unfortunately, LLVM isn't clean for this warning.
-                            "-Wno-unused-parameter",
+                            # Don't warn on external code as we can't
+                            # necessarily patch it easily.
+                            "--system-header-prefix=external/",
```
Comment (jonmeow) [inline bazel/cc_toolchains/clang_cc_toolchain_config.bzl:138]: Is this necessary with the bazelrc change?

File: explorer/syntax/parse_and_lex_context.h [hunk 2]
```diff
@@ -55,9 +55,9 @@ class ParseAndLexContext {
 }  // namespace Carbon
 
 // Gives flex the yylex prototype we want.
-#define YY_DECL                                                         \
-  auto yylex(Carbon::Nonnull<Carbon::Arena*> arena, yyscan_t yyscanner, \
-             Carbon::ParseAndLexContext& context)                       \
+#define YY_DECL                                                             \
+  auto yylex(Carbon::Nonnull<Carbon::Arena*> /*arena*/, yyscan_t yyscanner, \
```
Comment (jonmeow) [inline explorer/syntax/parse_and_lex_context.h:59]: This change seems odd to me, it does get used...

**Review by jonmeow — APPROVED** [review]

File: bazel/cc_toolchains/clang_cc_toolchain_config.bzl (hunk 1 above)
Comment (chandlerc) [reply bazel/cc_toolchains/clang_cc_toolchain_config.bzl:138]: Yeah. This covers headers from these directories being included into Carbon code.

File: explorer/syntax/parse_and_lex_context.h (hunk 2 above)
Comment (chandlerc) [reply explorer/syntax/parse_and_lex_context.h:59]: I mean, if you uncomment it, Clang warns about it in one context. I didn't investigate further when commenting it out fixed the compile error.

This is declaring the *lexer* which I don't think uses this, but needs to have the same signature as the parser?

**Comment by chandlerc** [comment]
(squashing and merging as discussed so that the action runs can start up that will populate the build cache with new artifacts)

**Pull request merged**